scarlet/source.py

```diff
@@ -1,16 +1,125 @@
 from functools import partial
 
 from .initialization import *
-from .constraint import PositivityConstraint, MonotonicityConstraint, SymmetryConstraint, L0Constraint
+from .constraint import (
+    PositivityConstraint,
+    MonotonicityConstraint,
+    SymmetryConstraint,
+    L0Constraint,
+)
 from .constraint import NormalizationConstraint, ConstraintChain, CenterOnConstraint
 from .parameter import Parameter, relative_step
-from .component import ComponentTree, FunctionComponent, FactorizedComponent
+from .component import ComponentTree, Factor, FactorizedComponent
 from .bbox import Box
-from .wavelet import Starlet, mad_wavelet
+from .wavelet import Starlet
+from .psf import PSF
+from . import fft
+from . import interpolation
 
 # make sure that import * above doesn't import its own stock numpy
 import autograd.numpy as np
 
+
+class Spectrum(Factor):
+    pass
+
+
+class TabulatedSpectrum(Spectrum):
+    def __init__(self, spectrum, bbox=None):
+        if isinstance(spectrum, Parameter):
+            assert spectrum.name == "spectrum"
+        else:
+            constraint = PositivityConstraint()
+            step = partial(relative_step, factor=1e-2)
+            spectrum = Parameter(
+                spectrum, name="spectrum", step=step, constraint=constraint
+            )
+
+        if bbox is None:
+            self._bbox = Box(spectrum.shape)
+        else:
+            assert bbox.shape == spectrum.shape
+            self._bbox = bbox
+
+        super().__init__(spectrum)
+
+    def get_model(self, *parameters):
+        spectrum = self.parameters[0]
+        for p in parameters:
+            if p._value.name == "spectrum":
+                spectrum = p
+        return spectrum
+
+
+class Morphology(Factor):
+    @property
+    def center(self):
+        return None
+
+
+class ImageMorphology(Morphology):
```

Comment:
Yes, it's a little more code to write initially, but no, it's less complicated. We had this pretty confusing mess of initialization, generation of some parameters, and then setting the component. The new API streamlines that considerably. Moreover, it makes the factors reusable, so folks don't have to copy&paste code around.

Reply:
Yes, that I like, but instead now, we pretty much have to copy-paste the code for sources as they almost always do the same. I wonder if there is not a way to automate this.

Reply:
I take this as a request for docstrings. The other is a conversation about how we deal with custom source code.

Reply:
Yes..

Reply:
What about putting priors on the factors, or altering the step sizes, or warm-starting parameters with previous gradients. All of the could be done by exposing these features to the API. I've opted for accepting a fully formed `Parameter` or creating one with default settings. I think this covered all bases and is in line with the principle that you need to generate parameters and their initialization together to make sense.

Reply:
But that isn't the way that it works in your PR. You still give the `ExtendedSourceMorphology` class three parameters that are used to build the constraint, it's only consistent because you set them in your source class.

To truly make sure the constraints and initialization are consistent we should get rid of the `ExtendedSourceMorphology` class and add a `constrains` parameter to `ImageMorphology`. Then in `ExtendedSource` you can build the constraints, which guarantees that they are consistent with the initialization and then just use `ImageMorphology` with the constraints generated in `ExtendedSource`.

Reply:
Fair. But this way reduces the copy/pasting for `MultiComponentSource`, which would otherwise have to go through the same logic again.

Reply:
If you just add `sed`, `morph`, and `bbox` arguments to `ExtendedSource` then you can just make the `children` of `MultiComponentSource` each `ExtendedSource`s with the initialization from `MultiComponentSource`. But even if you don't, I'd rather us have this duplicate code in `source` than make more work for ourselves and other users by also having to create a `Morphology` class.

Reply:
I don't follow